include/seqan3/io/detail/safe_filesystem_entry.hpp

```diff
@@ -0,0 +1,97 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::safe_filesystem_entry.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <cassert>
+#include <system_error>
+
+#include <seqan3/core/platform.hpp>
+#include <seqan3/std/filesystem>
+
+namespace seqan3::detail
+{
+
+/*!\brief A safe guard to manage a filesystem entry, e.g. a file or a directory.
+ * \ingroup io
+ *
+ * \details
+ *
+ * This raii-wrapper class allows for a safe removal of a created filesystem entry such as a directory or file.
+ * This wrapper class assumes owning semantics. It is not copy-constructible or copy-assignable. In order to
+ * prevent misuse also the default constructor is deleted.
+ *
+ * The following example demonstrates the use case.
+ *
+ * \include test/snippet/io/detail/safe_filesystem_entry_snippet.cpp
+ */
+class safe_filesystem_entry
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    safe_filesystem_entry() = delete;                                             //!< Deleted.
+    safe_filesystem_entry(safe_filesystem_entry const &) = delete;                //!< Deleted.
+    safe_filesystem_entry(safe_filesystem_entry &&) = default;                    //!< Defaulted.
+    safe_filesystem_entry & operator=(safe_filesystem_entry const &) = delete;    //!< Deleted.
+    safe_filesystem_entry & operator=(safe_filesystem_entry &&) = default;        //!< Defaulted.
+
+    /*!\brief Constructs the safe guard from a std::filesystem::path.
+     * \param p The path pointing to a filesystem entry.
+     */
+    safe_filesystem_entry(std::filesystem::path p) : entry(std::move(p))
+    {}
+
+    //!\brief Calls std::filesystem::remove_all on the wrapped entry.
+    ~safe_filesystem_entry()
+    {
+        std::error_code ec;
+        std::filesystem::remove_all(entry, ec);
+
+        assert(!static_cast<bool>(ec));
+    }
+    //!\}
+
+    /*!\brief Removes a file or empty directory.
+     * \returns `true` if the  underlying path could be removed, otherwise `false`.
```

Comment:
This is not correct. According to the documentation it only returns false if the file didn't exist, because you are using the throwing-interface.

This is another example of were we (unfortunately also the standard to some degree) mixes return values and exceptions for error-handling. I would suggest uses the error_code interface and just ignoring the error code.

Reply:
Ok, I see that the docs could be updated but I disagree on the second part. I strongly advice not to ignore the warnings. If an exception is thrown, then because something severe happened. Finding the cause of this is very hard without an exception. We don't do embedded coding, which is the actually the main reason for also offering the error code versions.

Reply:
But the thing that is incorrect in the docs is exactly the thing everyone will get wrong when they use the function. Because there is `std::remove_all` that throws, `std::remove_all` that gets error_code out-parameter and now there is seqan remove_all that returns bool, but also sometimes throws. It's a little messy...

Reply:
I did have a wrong return type for remove_all. That needs to be fixed yes. I can also add a second overload with error_code. But otherwise they just wrap the filesystem call nothing more.